### ISSUE TITLE
Segment redraw bug in Firefox?

### ISSUE THREAD

**Issue by ffxsam** [bbc/peaks.js#358]
Has anyone else ever seen anything like this in Firefox? I can't replicate it at all in Chrome nor Safari.

I'm not sure if this is a strange bug in Peaks.js, or a Firefox canvas bug.

https://user-images.githubusercontent.com/12532733/105121044-09739d00-5a99-11eb-8feb-9e922b164cc7.mp4

**Comment by dutzi** [comment]
Hey @ffxsam, did you manage to find the source for this?

**Comment by ffxsam** [comment]
@dutzi Unfortunately not, still an issue here - last time I checked, at least.

**Comment by dutzi** [comment]
@ffxsam, I think I found it, changing [this block](https://github.com/bbc/peaks.js/blob/master/src/waveform-shape.js#L102-L110) to:

```js
this._drawWaveform(
  context,
  this._view.getWaveformData(),
  Math.round(frameOffset),
  Math.round(this._segment ? this._view.timeToPixels(this._segment.startTime) : frameOffset),
  Math.floor(this._segment ? this._view.timeToPixels(this._segment.endTime)   : frameOffset + width),
  width,
  height
);
```

Wrapping the values with `round` and `floor` seems to fix this.

**Comment by chrisn** [comment]
I haven't observed this issue myself yet. I wonder why those changes fix it, do we have non-integer values somewhere?

**Comment by dutzi** [comment]
When playing around with the zoom factor `frameOffset` would sometimes be n+0.5 (148.5, for example) 🤷‍♂️

**Comment by chrisn** [comment]
I think that the fraction values may originate from mousedown and mousemove events. I have just pushed a change that calls `Math.floor()` on those values, which hopefully fixes this issue.

**Comment by dutzi** [comment]
I think it happens when rendering peaks in a 4K or Retina displays. It might be that in such setup, dots cover 4 (or more) pixels, 2 per each axis, that would also explain why I only saw float that look like x.5 (and not x.72, for example).

**Comment by chrisn** [comment]
Yes, I think you're right - although I haven't seen non-integer values on my Windows 10 laptop with 4K display. `window.devicePixelRatio` reports 2.25. Would you mind testing the change in https://github.com/bbc/peaks.js/commit/77440c9cccda638e0b20ccc09955515d2030c870 to see if it works for you?

**Comment by ffxsam** [comment]
I noticed there's some color bleed at the edges of the waveforms. I'm wondering if this has something to do with it, like one layer is clipping over the other one for some reason?

![wave-edges](https://user-images.githubusercontent.com/12532733/117585061-0519d100-b0d6-11eb-9288-068a8982236c.png)

**Comment by chrisn** [comment]
I haven't been able to reproduce this locally, in Firefox or Chrome. Which OS and browser version are you using?

**Comment by ffxsam** [comment]
macOS 10.15.7, Firefox 88.0.1. I believe this only occurs in Firefox.

**Comment by ffxsam** [comment]
I'll dig into this when I get a chance and let you know.
